### PR TITLE
Sort Preserving RepartitionExec implementation

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/repartition/sort_preserving_repartition.rs [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+        }
+        let mut stream = streaming_merge(
+            input_streams,
+            schema,
+            &sort_exprs,
+            BaselineMetrics::new(&metrics, partition),
+            context.session_config().batch_size(),
+        )?;
+        // let mut stream = input.execute(i, context)?;
+        timer.done();
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/repartition/sort_preserving_repartition.rs:309]: This section is what is different than the existing `RepartitionExec`

File: datafusion/core/src/physical_plan/repartition/mod.rs [hunk 2]
```diff
@@ -681,6 +552,132 @@ impl RecordBatchStream for RepartitionStream {
     }
 }
 
+/// Pulls data from the specified input plan, feeding it to the
+/// output partitions based on the desired partitioning
+///
+/// i is the input partition index
+///
+/// txs hold the output sending channels for each output partition
+async fn pull_from_input(
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/repartition/mod.rs:561]: There is no change here, function is just removed from self to outside (We now can use this for SortPreservingRepartitionExec).

File: datafusion/core/src/physical_plan/repartition/mod.rs [hunk 3]
```diff
@@ -681,6 +552,132 @@ impl RecordBatchStream for RepartitionStream {
...
+
+    Ok(())
+}
+
+/// Waits for `input_task` which is consuming one of the inputs to
+/// complete. Upon each successful completion, sends a `None` to
+/// each of the output tx channels to signal one of the inputs is
+/// complete. Upon error, propagates the errors to all output tx
+/// channels.
+async fn wait_for_task(
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_plan/repartition/mod.rs:642]: There is no change here, function is just removed from self to outside (Now, We can use this for SortPreservingRepartitionExec).

**Review by ozankabak — APPROVED** [review]
LGTM. Let's write a PR body and send upstream to get community feedback on our approach.

FYI, fuzz tests live in a separate place now -- let's not forget to add them there from now on (I already fixed it for this PR).

**Comment by ozankabak** [comment]
Merged upstream, closing.

**Pull request closed without merging**